include/tlapack/plugins/starpu.hpp

```diff
@@ -0,0 +1,927 @@
+            {
+                return sqrt(T(x));
+            }
+
+            // Display value in ostream
+
+            constexpr friend std::ostream& operator<<(std::ostream& out,
+                                                      const data& x)
+            {
+                return out << T(x);
+            }
+
+           private:
+            /**
+             * @brief Applies an operation and assigns
+             *
+             * Operations: +, -, *, /
+             *
+             * @tparam op  Operation
+             * @param x   Second operand
+             * @return constexpr data&  Reference to the result
+             */
+            template <Operation op>
+            constexpr data& operate_and_assign(const data& x) noexcept
+            {
+                constexpr starpu_data_access_mode mode =
+                    (op == Operation::Assign) ? STARPU_W : STARPU_RW;
+
+                // Create the codelet
+                struct starpu_codelet cl;
+                starpu_codelet_init(&cl);
+                cl.cpu_funcs[0] = cpu::data_op_data<T, op>;
+                cl.nbuffers = 2;
+                cl.modes[0] = mode;
+                cl.modes[1] = STARPU_R;
+                cl.name = (to_string(op) + "_data_" + typeid(T).name()).c_str();
+
+                struct starpu_task* task = starpu_task_create();
+                task->cl = &cl;
+                task->handles[0] = this->handle;
+                task->handles[1] = x.handle;
+                task->synchronous = true;
+                const int ret = starpu_task_submit(task);
+
+                // const int ret = starpu_task_insert(
+                //     &cl, mode, this->handle, STARPU_R, x.handle,
+                //     STARPU_TASK_SYNCHRONOUS, true, 0);
+
+                STARPU_CHECK_RETURN_VALUE(ret, "starpu_task_submit");
+
+                return *this;
+            }
+
+            /**
+             * @brief Applies an operation and assigns
+             *
+             * Operations: +, -, *, /
+             *
+             * @tparam op  Operation
+             * @param x   Second operand
+             * @return constexpr data&  Reference to the result
+             */
+            template <Operation op>
+            constexpr data& operate_and_assign(const T& x) noexcept
+            {
+                constexpr starpu_data_access_mode mode =
+                    (op == Operation::Assign) ? STARPU_W : STARPU_RW;
+
+                // Create the codelet
+                struct starpu_codelet cl;
+                starpu_codelet_init(&cl);
+                cl.cpu_funcs[0] = cpu::data_op_value<T, op>;
+                cl.nbuffers = 1;
+                cl.modes[0] = mode;
+                cl.name =
+                    (to_string(op) + "_value_" + typeid(T).name()).c_str();
+
+                // Allocate space for x
+                void* x_copy;
+                const int info = starpu_malloc(&x_copy, sizeof(T));
```

Comment:
you do not need to starpu_malloc (i.e. pin) it: since it will be passed as is, without any cuda transfer or something, you can just malloc it.